test/e2e/funcs/feature.go

```diff
@@ -348,23 +348,44 @@ func ResourceHasConditionWithin(d time.Duration, o k8s.Object, cds ...xpv1.Condi
 		desired := strings.Join(reasons, ", ")
 
 		t.Logf("Waiting %s for %s to become %s...", d, identifier(o), desired)
+		ogReport := make(map[string]bool)
 		old := make([]xpv1.Condition, len(cds))
 		match := func(o k8s.Object) bool {
 			u := asUnstructured(o)
 			s := xpv1.ConditionedStatus{}
 			_ = fieldpath.Pave(u.Object).GetValueInto("status", &s)
 
 			for i, want := range cds {
+				// Update the wanted observed generation to the latest object generation.
+				want.ObservedGeneration = u.GetGeneration()
+
 				got := s.GetCondition(want.Type)
-				if !got.Equal(old[i]) {
+
+				// Until https://github.com/crossplane/crossplane/issues/6420 is resolved, crossplane will be in a
+				// transition period. A condition with an observedGeneration of zero means it is not yet being
+				// propagated when setting the conditions. To help that transition, we will move the generation forward
+				// ONLY if it is zero. But we will also log this fact to find it in the logs.
+				if got.ObservedGeneration == 0 {
+					got.ObservedGeneration = u.GetGeneration()
+					key := fmt.Sprintf("%s[%s]", u.GetKind(), got.Type)
+					if !ogReport[key] {
+						ogReport[key] = true
+						t.Logf("crossplane#6420: Warning, an unset observedGeneration was atrifically updated for %s.status.conditions[%s]", u.GetKind(), got.Type)
```

Comment:
```suggestion
						t.Logf("crossplane#6420: Warning, an unset observedGeneration was artificially updated for %s.status.conditions[%s]", u.GetKind(), got.Type)
```

I'd prefer to log the actual issue URL here. I'm not sure everyone would recognize this syntax as referencing a GitHub issue.